FirebaseStorage/Sources/Internal/StoragePath.swift

```diff
@@ -0,0 +1,209 @@
+        "of gs://, http://, or https://")
+    }
+  }
+
+  /**
+   * Parses a gs://bucket/path/to/object URI into a GCS path.
+   * @param aURIString gs:// URI which is parsed into a path.
+   * @return Returns an instance of StoragePath or nil if one can't be created.
+   * @throws Throws an exception if the string is not a valid gs:// URI.
+   */
+  internal static func path(GSURI aURIString: String) throws -> StoragePath {
+    if aURIString.starts(with: "gs://") {
+      let bucketObject = aURIString.dropFirst("gs://".count)
+      if bucketObject.contains("/") {
+        let splitStringArray = bucketObject.split(separator: "/", maxSplits: 1).map(String.init)
+        let object = splitStringArray.count == 2 ? splitStringArray[1] : nil
+        return StoragePath(with: splitStringArray[0], object: object)
+      } else if bucketObject.count > 0 {
+        return StoragePath(with: String(bucketObject))
+      }
+    }
+    throw StoragePathError
+      .storagePathError("Internal error: URI must be in the form of " +
+        "gs://<bucket>/<path/to/object>")
+  }
+
+  /**
+   * Parses a http[s]://firebasestorage.googleapis.com/v0/b/bucket/o/path/to/object...?token=<12345>
+   * URL into a GCS path.
+   * @param aURLString http[s]:// URL which is parsed into a path.
+   * string which is parsed into a path.
+   * @return Returns an instance of StoragePath or nil if one can't be created.
+   * @throws Throws an exception if the string is not a valid http[s]:// URL.
+   */
+  private static func path(HTTPURL aURLString: String) throws -> StoragePath {
+    let httpsURL = URL(string: aURLString)
+    let pathComponents = httpsURL?.pathComponents
+    guard let pathComponents = pathComponents,
+          pathComponents.count >= 4,
+          pathComponents[1] == "v0",
+          pathComponents[2] == "b" else {
+      throw StoragePathError.storagePathError("Internal error: URL must be in the form of " +
+        "http[s]://<host>/v0/b/<bucket>/o/<path/to/object>[?token=signed_url_params]")
+    }
+    let bucketName = pathComponents[3]
+
+    guard pathComponents.count > 4 else {
+      return StoragePath(with: bucketName)
+    }
+    // Construct object name
+    var objectName = pathComponents[5]
+    for i in 6 ..< pathComponents.count {
+      objectName = "\(objectName)/\(pathComponents[i])"
+    }
+    return StoragePath(with: bucketName, object: objectName)
+  }
+
+  // Removes leading and trailing slashes, and compresses multiple slashes
+  // to create a canonical representation.
+  // Example: /foo//bar///baz//// -> foo/bar/baz
+  private static func standardizedPathForString(_ string: String?) -> String? {
+    guard let string = string else {
+      return nil
+    }
+    var output = string
+    while true {
+      let newOutput = output.replacingOccurrences(of: "//", with: "/")
+      if newOutput == output {
+        break
+      }
+      output = newOutput
+    }
```

Comment:
Kind of an obtuse way of writing this, but you could replace this `while true` loop with a reduce.

```swift
let slash: Character = "/"
output = string.reduce(into: "") { partialResult, next in
  guard partialResult.last.map({ $0 == slash }) ?? false else {
    return partialResult.append(next)
  }
  return next == slash ? () : partialResult.append(next) // this line here is especially code-golfy
}
```

Also, note that the method signature could be simplified from `String? -> String?` to `String -> String`, and we can instead lean on language constructs for dealing with optional values outside the function implementation.

Reply:
Fixed the signature. The other part is interesting, but at least to me is significantly more obtuse, so I'll skip/.